request-manager/request/resumer.go

```diff
@@ -0,0 +1,620 @@
+	// Start a transaction.
+	txn, err := conn.BeginTx(context.TODO(), nil)
+	if err != nil {
+		return err
+	}
+	defer txn.Rollback()
+
+	err = r.updateRequestWithTxn(request, curState, txn)
+	if err != nil {
+		return err
+	}
+
+	return txn.Commit()
+}
+
+// updateRequestWithTxn updates the State and JR host as set in the given request,
+// using the provided db transaction. The request is updated only if its current
+// state in the db matches the state provided.
+func (r *resumer) updateRequestWithTxn(request proto.Request, curState byte, txn *sql.Tx) error {
+	// If JobRunnerHost is empty, we want to set the db field to NULL.
+	var jrHost sql.NullString
+	if request.JobRunnerHost == "" {
+		jrHost = sql.NullString{}
+	} else {
+		jrHost = sql.NullString{
+			String: request.JobRunnerHost,
+			Valid:  true,
+		}
+	}
+
+	// Update the 'state' and 'jr_host' fields only.
+	q := "UPDATE requests SET state = ?, jr_host = ? WHERE request_id = ? AND state = ?"
+	res, err := txn.Exec(q, request.State, jrHost, request.Id, curState)
+	if err != nil {
+		return err
+	}
+
+	cnt, err := res.RowsAffected()
+	if err != nil {
+		return err
+	}
+
+	switch cnt {
+	case 0:
+		// Either the request's current state != curState, or no request with the
+		// id given exists.
+		return db.ErrNotUpdated
+	case 1:
+		return nil
+	default:
+		// This should be impossible since we specify the primary key (request id)
+		// in the WHERE clause of the update.
+		return db.ErrMultipleUpdated
+	}
+}
+
+// deleteSJC removes an SJC from the db. The RM needs to have claimed the
+// SJC in order to delete it.
+func (r *resumer) deleteSJC(requestId string, conn *sql.Conn) error {
+	// Start a transaction.
+	txn, err := conn.BeginTx(context.TODO(), nil)
+	if err != nil {
+		return err
+	}
+	defer txn.Rollback()
+
+	q := "DELETE FROM suspended_job_chains WHERE request_id = ? AND rm_host = ?"
+	result, err := txn.Exec(q,
+		requestId,
+		r.host)
+	if err != nil {
+		return err
+	}
+
+	count, err := result.RowsAffected()
+	if err != nil {
+		return err
+	}
+
+	switch count {
+	case 0:
+		return fmt.Errorf("cannot find SJC to delete - may not be claimed by RM")
+	case 1: // Success
+		log.Infof("deleted sjc %s", requestId)
+		return txn.Commit()
+	default:
+		// This should be impossible since we specify the primary key (request id)
+		// in the WHERE clause of the update.
+		return db.ErrMultipleUpdated
+	}
+}
+
+// Claim that this RM will resume the SJC corresponding to the request id.
+// Returns true if SJC is successfully claimed (this RM should resume the SJC);
+// returns false if failed to claim SJC (another RM is resuming the SJC).
+// claimSJC is used to make sure that only one RM is trying to resume a specific
+// SJC at any given time.
+func (r *resumer) claimSJC(requestId string, conn *sql.Conn) (bool, error) {
+	// Start a transaction
+	ctx := context.TODO()
+	txn, err := conn.BeginTx(ctx, nil)
```

Comment:
Don't need a txn in this case. A single statement with auto-commit (enabled by default) is an implicit transaction.